eat: add express endpoints for lightning invoice

007harshmahajan opened the pull request:
see commit details 

TICKET: BTC-1836

<!--
# Please be aware of the following when making your pull request:

## Description

Please include a summary of your proposed changes and which issue is being addressed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Issue Number

Internal Users - Please include the related internal tracking number (e.g. BG-000000).
External Users - Please link to any relevant github issues as necessary.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My code compiles correctly for both Node and Browser environments
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [ ] The ticket or github issue was included in the commit message as a reference
- [ ] I have made corresponding changes to the documentation and on any new/updated functions and/or methods - [jsdoc](https://jsdoc.app/)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
-->

ericli-bitgo commented on modules/sdk-coin-lnbtc/src/register.ts:
```diff
@@ -0,0 +1,8 @@
+import { Tlnbtc } from './tbtc';
+
+export const register = (sdk: BitGoBase): void => {
+  sdk.register('btc', Lnbtc.createInstance);
```
```suggestion
  sdk.register('lnbtc', Lnbtc.createInstance);
```

ericli-bitgo commented on modules/sdk-coin-lnbtc/src/register.ts:
```diff
@@ -0,0 +1,8 @@
+
+export const register = (sdk: BitGoBase): void => {
+  sdk.register('btc', Lnbtc.createInstance);
+  sdk.register('tbtc', Tlnbtc.createInstance);
```
```suggestion
  sdk.register('tlnbtc', Tlnbtc.createInstance);
```

ericli-bitgo reviewed (changes requested):
Add register.ts to lnbtc/index.ts

lcovar committed:
feat: add proxyquire dev import

This is needed since we need to somehow stub the getLighhtningWallet
method within the express unit tests but this method lives in another
module. Proxyquire was the only way I could figure this out.

Ticket: BTC-1836

TICKET: BTC-1836

lcovar committed:
feat: add express endpoints for lightning invoice

Add
- create invoice
- pay invoice

Ticket: BTC-1836

TICKET: BTC-1836

lcovar commented on modules/abstract-lightning/src/wallet/lightning.ts:
```diff
@@ -104,7 +104,7 @@ export class SelfCustodialLightningWallet implements ILightningWallet {
     const createInvoiceResponse = await this.wallet.bitgo
       .post(this.wallet.baseCoin.url(`/wallet/${this.wallet.id()}/lightning/invoice`))
-      .send(CreateInvoiceBody.encode(params))
+      .send(t.exact(CreateInvoiceBody).encode(params))
```
this should have been a different commit but with this we won't send any extra params to WP if client accidentally passes params not in the codec.

saravanan7mani commented on modules/abstract-lightning/src/codecs/api/payment.ts:
```diff
@@ -88,7 +88,7 @@ export type PaymentQuery = t.TypeOf<typeof PaymentQuery>;
 
 export const SubmitPaymentParams = t.intersection([
   LightningPaymentRequest,
```
Let's add `passphrase` here.

saravanan7mani commented on modules/express/src/lightning/lightningInvoiceRoutes.ts:
```diff
@@ -0,0 +1,44 @@
+
+export async function handlePayLightningInvoice(req: express.Request): Promise<any> {
+  const bitgo = req.bitgo;
+  const { passphrase } = req.body;
+  if (passphrase === undefined) {
+    throw new ApiResponseError('Missing wallet passphrase', 400);
+  }
```
lets remove it and use codec

saravanan7mani commented on modules/express/src/lightning/lightningInvoiceRoutes.ts:
```diff
@@ -0,0 +1,44 @@
+    const wallet = await coin.wallets().get({ id: req.params.id });
+    const lightningWallet = getLightningWallet(wallet);
+
+    return await lightningWallet.payInvoice(params, passphrase);
```
Lets change params to pass passphrase as well.

`return await lightningWallet.payInvoice(params);`

saravanan7mani commented on modules/express/src/clientRoutes.ts:
```diff
@@ -1704,4 +1705,16 @@ export function setupLightningRoutes(app: express.Application, config: Config):
     promiseWrapper(handleUnlockLightningWallet)
   );
   app.get('/api/v2/:coin/wallet/:id/state', prepareBitGo(config), promiseWrapper(handleGetLightningWalletState));
+  app.post(
+    '/api/v2/:coin/wallet/:id/lightning/invoice',
+    parseBody,
+    prepareBitGo(config),
+    promiseWrapper(handleCreateLightningInvoice)
+  );
+  app.post(
+    '/api/v2/:coin/wallet/:id/lightning/pay',
+    parseBody,
+    prepareBitGo(config),
+    promiseWrapper(handlePayLightningInvoice)
+  );
```
`setupLightningRoutes` is supposed contain APIs those work with signer LND. Express without signer LND config also should be allowed to work with invoice APIs. They can be added directly within `setupAPIRoutes`.
